config/include.go

```diff
@@ -133,6 +225,174 @@ func (targetConfig *TerragruntConfig) Merge(sourceConfig *TerragruntConfig, terr
+	// Merge simple attributes first
+	if sourceConfig.DownloadDir != "" {
+		targetConfig.DownloadDir = sourceConfig.DownloadDir
+	}
+
+	if sourceConfig.IamRole != "" {
+		targetConfig.IamRole = sourceConfig.IamRole
+	}
+
+	if sourceConfig.IamAssumeRoleDuration != nil {
+		targetConfig.IamAssumeRoleDuration = sourceConfig.IamAssumeRoleDuration
+	}
+
+	if sourceConfig.TerraformVersionConstraint != "" {
+		targetConfig.TerraformVersionConstraint = sourceConfig.TerraformVersionConstraint
+	}
+
+	if sourceConfig.TerraformBinary != "" {
+		targetConfig.TerraformBinary = sourceConfig.TerraformBinary
+	}
+
+	if sourceConfig.PreventDestroy != nil {
+		targetConfig.PreventDestroy = sourceConfig.PreventDestroy
+	}
+
+	if sourceConfig.RetryMaxAttempts != nil {
+		targetConfig.RetryMaxAttempts = sourceConfig.RetryMaxAttempts
+	}
+
+	if sourceConfig.RetrySleepIntervalSec != nil {
+		targetConfig.RetrySleepIntervalSec = sourceConfig.RetrySleepIntervalSec
+	}
+
+	if sourceConfig.TerragruntVersionConstraint != "" {
+		targetConfig.TerragruntVersionConstraint = sourceConfig.TerragruntVersionConstraint
+	}
+
+	// Skip has to be set specifically in each file that should be skipped
+	targetConfig.Skip = sourceConfig.Skip
+
+	// Handle list attributes by concatenatenation
+	if sourceConfig.Dependencies != nil {
+		resultModuleDependencies := &ModuleDependencies{}
+		if targetConfig.Dependencies != nil {
+			resultModuleDependencies.Paths = targetConfig.Dependencies.Paths
+		}
+		resultModuleDependencies.Paths = append(resultModuleDependencies.Paths, sourceConfig.Dependencies.Paths...)
+		targetConfig.Dependencies = resultModuleDependencies
+	}
+
+	if sourceConfig.RetryableErrors != nil {
+		targetConfig.RetryableErrors = append(targetConfig.RetryableErrors, sourceConfig.RetryableErrors...)
+	}
+
+	// Handle complex structs by recursively merging the structs together
+	if sourceConfig.Terraform != nil {
+		if targetConfig.Terraform == nil {
+			targetConfig.Terraform = sourceConfig.Terraform
+		} else {
+			if sourceConfig.Terraform.Source != nil {
+				targetConfig.Terraform.Source = sourceConfig.Terraform.Source
+			}
+			mergeExtraArgs(terragruntOptions, sourceConfig.Terraform.ExtraArgs, &targetConfig.Terraform.ExtraArgs)
+
+			mergeHooks(terragruntOptions, sourceConfig.Terraform.BeforeHooks, &targetConfig.Terraform.BeforeHooks)
+			mergeHooks(terragruntOptions, sourceConfig.Terraform.AfterHooks, &targetConfig.Terraform.AfterHooks)
+		}
+	}
+
+	if sourceConfig.Inputs != nil {
+		mergedInputs, err := deepMergeInputs(sourceConfig.Inputs, targetConfig.Inputs)
+		if err != nil {
+			return err
+		}
+		targetConfig.Inputs = mergedInputs
+	}
+
+	// MAINTAINER'S NOTE: The following structs cannot be deep merged due to an implementation detail (they do not
+	// support nil attributes, so we can't determine if an attribute was intentionally set, or was defaulted from
+	// unspecified - this is especially problematic for bool attributes).
```

Comment:
Ah, I see... So in the future, we'd need to update these to support `nil`?

Reply:
Yea because we can't tell if a `false` bool attribute was intentionally set in the code in the child `terragrunt.hcl` config, or if it was because it was unspecified. If unspecified, then we want to inherit from the parent, but if it was intentionally set, then it needs to override the parent. Since the resulting behavior is different, it's important that we can differentiate the two cases.

I decided to avoid addressing this for now in this PR since it was already getting big, but we will definitely want to address this as it feels limiting to not be able to deep merge an important block like `remote_state`.